Broaden type annotation for verbose_name(_plural) to accept lazystr.

PIG208 opened the pull request:
# I have made things!

<!--
Hi, thanks for submitting a Pull Request. We appreciate it.

Please, fill in all the required information
to make our review and merging processes easier.

Cheers!
-->
There might be other use cases of `lazystr` that need to be considered. It needs to be verified that `Promise` objects are actually expected for the callee. Right now we are following a conservative approach to first mark ones that do commonly accept `Promise`.

## Related issues

- Closes #1137
- Refs #1137

<!--
Mark what issues this Pull Request closes or references.

Format is:
- Closes #issue-number
- Refs #issue-number

Example. Refs #0
Documentation: https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/
-->

<!--
If you have any feedback, just write it here.

It can be whatever you want!
-->

PIG208 committed:
Broaden type annotation for verbose_name(_plural) to accept lazystr.

Fixes #1137.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

sobolevn reviewed (commented):
Is there any way to make `str` and `Promise` compatible?

PIG208 committed:
Broaden type annotation for help_text to accept lazystr.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Broaden type annotation for ValidationError to accept lazystr.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
Broaden type annotation for label to accept lazystr.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

flaeppe commented on django-stubs/utils/functional.pyi:
```diff
@@ -43,6 +43,7 @@ class _StrPromise(Promise, Sequence[str]):
     # Mypy requires this for the attribute hook to take effect
     def __getattribute__(self, __name: str) -> Any: ...
 
+_StrOrPromise = Union[str, _StrPromise]
```
Maybe `_StrPromise` should be public? Since it'll be exposed to users, right? 

So that one can do `str | StrPromise` if passing around/accepting `gettext_lazy` values

```suggestion
_StrOrPromise = Union[str, StrPromise]
```

flaeppe reviewed (commented):
I can't come up with any better idea really. But I'm curious if these changes could result in users prematurely evaluating lazy strings.

But I guess those cases could be mitigated by patching the stubs.

Suppose we'll notice in time

PIG208 committed:
Add StrPromise and StrOrPromise aliases to django_stubs_ext.

We make StrPromise and StrOrPromise available via django_stubs_ext so
that conditional imports with TYPE_CHECKING is not required.
These aliases fall back to Promise or Union[str, Promise]
when not TYPE_CHECKING.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

andersk commented on django_stubs_ext/django_stubs_ext/aliases.py:
```diff
@@ -2,9 +2,15 @@
 
     ValuesQuerySet = _QuerySet[_T, _Row]
+    StrOrPromise = StrOrPromise
+    StrPromise = StrPromise
```
This isn’t [how re-exports work](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport) in mypy.

But I think your initial inclination to use leading underscores in `django-stubs/utils/functional.pyi` was correct, because these don’t exist in `django.utils.functional` at runtime. We should use underscores in `django-stubs` and re-export with no underscores here in `django_stubs_ext`.

flaeppe commented on django_stubs_ext/django_stubs_ext/aliases.py:
```diff
@@ -2,9 +2,15 @@
 
 if typing.TYPE_CHECKING:
     from django.db.models.query import _T, _QuerySet, _Row
+    from django.utils.functional import _StrOrPromise, _StrPromise
```
It's fair to have this re-exported in aliases (I think it should work?)

```suggestion
    from django.utils.functional import _StrOrPromise as StrOrPromise, _StrPromise as StrPromise
```

PIG208 replied:
This fails the mypy check https://github.com/typeddjango/django-stubs/actions/runs/3076096523/jobs/4970014010. We will need to allow implicit_reexport for `django_stubs_ext.aliases` if we do it this way.

PIG208 replied:
Actually, never mind. Adding these aliases to `__all__` should fix the CI.

andersk replied:
@flaeppe Mypy is very specific about [what counts as a re-export](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport):

```python3
# This won't re-export the value
from foo import bar

# Neither will this
from foo import bar as bang

# This will re-export it as bar and allow other modules to import it
from foo import bar as bar

# This will also re-export bar
from foo import bar
__all__ = ['bar']
```

PIG208 replied:
If I understand correctly, mypy does not mention that this below is discouraged, right? Though we may as well define a type alias at this point.

```python
from foo import bar as bang
__all__ = ['bang']
```

flaeppe commented on django_stubs_ext/django_stubs_ext/aliases.py:
```diff
@@ -2,9 +2,15 @@
 
     ValuesQuerySet = _QuerySet[_T, _Row]
+    StrOrPromise = _StrOrPromise
+    StrPromise = _StrPromise
```
```suggestion
```

flaeppe commented on django_stubs_ext/django_stubs_ext/aliases.py:
```diff
@@ -2,9 +2,15 @@
+    StrPromise = _StrPromise
 else:
     from django.db.models.query import QuerySet
+    from django.utils.functional import Promise
```
```suggestion
    from django.utils.functional import Promise as StrPromise
```

flaeppe commented on django_stubs_ext/django_stubs_ext/aliases.py:
```diff
@@ -2,9 +2,15 @@
+    from django.utils.functional import Promise
 
     ValuesQuerySet = QuerySet
+    StrPromise = Promise
```
```suggestion
```

sobolevn reviewed (approved):
LGTM, any other comments? 

Thanks for working on this!

flaeppe reviewed (approved):
All good from my end
